File: LCCI/2-1.remove-duplicate-node-lcci/remove-duplicate-node-lcci.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

//超出时间限制
class Solution {
    public ListNode removeDuplicateNodes(ListNode head) {
        HashSet<Integer> set = new HashSet<Integer>();
        ListNode previous = null;
        while (head != null) {
            if (set.contains(head.val)) {
                previous.next = head.next;
            } else {
                set.add(head.val);
                previous = head;
            }
        }
        head = head.next;
        return head;
    }
}

File: LCCI/2-1.remove-duplicate-node-lcci/remove-duplicate-node-lcci.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeDuplicateNodes(head *ListNode) *ListNode {
    //考虑特殊情况
    if head == nil || head.Next == nil {
		return head
	}
	m := make(map[int]bool, 0)
	fake := new(ListNode)
	fake.Next = head
	cur := head
	pre := fake

	for cur != nil {
		if _, ok := m[cur.Val]; ok {
			cur = cur.Next
		} else {
			m[cur.Val] = true
            //删除重复的链表节点
			pre.Next = cur
			pre = pre.Next
			cur = cur.Next
		}
	}
	pre.Next = nil
	return head
}
